commons-impl/src/main/java/com/connexta/transformation/commons/api/impl/AbstractTransformationImpl.java

```diff
@@ -0,0 +1,297 @@
+
+  /**
+   * Instantiates a transformation based on the information provided by the specified pojo.
+   *
+   * @param pojo the pojo to initializes the transformation with
+   * @param clock the clock to use for retrieving wall and monotonic times
+   * @throws InvalidFieldException if an error occurs while trying to deserialize the pojo object
+   * @throws PersistenceException if an error occurs while deserializing the pojo object
+   */
+  public AbstractTransformationImpl(TransformationPojo pojo, Clock clock)
+      throws PersistenceException {
+    this(clock);
+    readFrom(pojo);
+  }
+
+  @VisibleForTesting
+  AbstractTransformationImpl(RequestInfo requestInfo, Instant startTime, Clock clock) {
+    this(clock);
+    this.requestInfo = requestInfo;
+    this.startTime = startTime;
+  }
+
+  @VisibleForTesting
+  AbstractTransformationImpl(Clock clock) {
+    super(AbstractTransformationImpl.PERSISTABLE_TYPE, null);
+    this.clock = clock;
+  }
+
+  @Override
+  public String getTransformId() {
+    return getId();
+  }
+
+  @Override
+  public RequestInfo getRequestInfo() {
+    return requestInfo;
+  }
+
+  @Override
+  public Instant getStartTime() {
+    return startTime;
+  }
+
+  @Override
+  public Duration getDuration() {
+    return Duration.between(
+        startTime, getCompletionTime().orElseGet(() -> Instant.ofEpochMilli(clock.wallTime())));
+  }
+
+  @Override
+  public boolean hasUnknowns() {
+    return hasUnknowns;
+  }
+
+  @Override
+  public Stream<MetadataTransformation> metadatas() {
+    return metadatas.values().stream().map(MetadataTransformation.class::cast);
+  }
+
+  @Override
+  public MetadataTransformation get(String type) throws TransformationNotFoundException {
+    final MetadataTransformation metadata = metadatas.get(type);
+
+    if (metadata == null) {
+      throw new TransformationNotFoundException(
+          "No [" + type + "] metadata found for transformation [" + getId() + "]");
+    }
+    return metadata;
+  }
+
+  /**
+   * Gets the clock to use for retrieving wall and monotonic times.
+   *
+   * @return the clock to use for retrieving wall and monotonic times
+   */
+  public Clock getClock() {
+    return clock;
+  }
+
+  /**
+   * Adds the given set of metadata to the set contained in this transformation.
+   *
+   * @param metadatas the set of metadata to combine with the contained set
+   */
+  public void add(Set<? super AbstractMetadataImpl> metadatas) {
+    add(metadatas.stream());
+  }
+
+  /**
+   * Adds the given metadata to the set contained in this transformation.
+   *
+   * @param metadatas the set of metadata to add to the contained set
+   */
+  public void add(AbstractMetadataImpl... metadatas) {
+    add(Stream.of(metadatas));
+  }
+
+  /**
+   * Adds the given metadata to the set contained in this transformation.
+   *
+   * @param metadatas the stream of metadata to add to the contained set
+   */
+  public void add(Stream<? super AbstractMetadataImpl> metadatas) {
```

Comment:
✏️ `VisibleForTesting` and package-private.

Reply:
same answer